File: binary_tree_paths.cpp
#include <cstdlib>
#include <vector>
#include <string>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    void dfs(TreeNode *root, string temp, vector<string> &vec) {
        if (!root)
            ;
        else {
            if (temp.empty())
                temp = std::to_string(root->val);
            else
                temp += ("->" + std::to_string(root->val));
            if (!root->left && !root->right)    // 若root结点为叶子结点，则push_back该路径
                vec.push_back(temp);
            else {
                dfs(root->left, temp, vec);
                dfs(root->right, temp, vec);
            }
        }
    }

    vector<string> binaryTreePaths(TreeNode* root) {
        vector<string> vec;
        string temp;    // temp初始化为空
        dfs(root, temp, vec);
        return vec;
    }
};

File: py/tree/BinaryTreePaths.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    def binaryTreePathsHelper(self, root, s, vec):
        if root.left == None and root.right == None:
            vec.append("{}{}".format(s, root.val))
        s += "{}->".format(root.val)
        if root.left != None:
            sLeft = s
            self.binaryTreePathsHelper(root.left, sLeft, vec)
        if root.right != None:
            sRight = s
            self.binaryTreePathsHelper(root.right, sRight, vec)
            
        
    def binaryTreePaths(self, root):
        if root == None:
            return None
        vec = []
        s = ""
        self.binaryTreePathsHelper(root, s, vec)
        return vec
